File: src/c3_otimes_c3/upb/parametrized_UPB.py
"""https://arxiv.org/abs/quant-ph/9908070

D. P. DiVincenzo, T. Mor, P. W. Shor, J. A. Smolin, B. M. Terhal,
"Unextendible Product Bases, Uncompletable Product Bases and Bound
Entanglement", Commun. Math. Phys. 238, 379 (2003), Section IV A. A
six-parameter family of unextendible product bases on C^3 (x) C^3 that
contains the Pyramid and Tiles UPBs (`pyramid_upb`, `tiles_upb`) as special
cases.
"""

from math import cos, sin, sqrt

import numpy as np
from utils import upb


def parametrized_basis(
    gamma_a: float,
    teta_a: float,
    phi_a: float,
    gamma_b: float,
    teta_b: float,
    phi_b: float,
) -> list[np.ndarray]:
    """The five product vectors of the six-parameter family of UPBs on C^3 (x) C^3.

    Reduces to the Pyramid UPB for phi_a = phi_b = 0 and teta_a = teta_b =
    gamma_a = gamma_b = arccos((sqrt(5) - 1) / 2), and to the Tiles UPB for
    phi_a = phi_b = 0 and teta_a = teta_b = gamma_a = gamma_b = 3*pi/4.

    Args:
        gamma_a: Alice's gamma angle. Requires cos(gamma_a) != 0 and
            sin(gamma_a) != 0 for the basis to be unextendible.
        teta_a: Alice's theta angle. Requires cos(teta_a) != 0 and
            sin(teta_a) != 0 for the basis to be unextendible.
        phi_a: Alice's phase angle.
        gamma_b: Bob's gamma angle, subject to the same restriction as gamma_a.
        teta_b: Bob's theta angle, subject to the same restriction as teta_a.
        phi_b: Bob's phase angle.

    Returns:
        list[np.ndarray]: the five normalized product vectors |a_i> (x) |b_i>
        forming the unextendible product basis, as column vectors.
    """
    N_a = sqrt(cos(gamma_a) ** 2 + sin(gamma_a) ** 2 * cos(teta_a) ** 2)
    N_b = sqrt(cos(gamma_b) ** 2 + sin(gamma_b) ** 2 * cos(teta_b) ** 2)

    a_0 = np.array(
        [
            1,
            0,
            0,
        ],
        dtype=np.complex128,
    )
    a_1 = np.array(
        [
            0,
            1,
            0,
        ],
        dtype=np.complex128,
    )
    a_2 = np.array(
        [
            cos(teta_a),
            0,
            sin(teta_a),
        ],
        dtype=np.complex128,
    )
    a_3 = np.array(
        [
            sin(gamma_a) * sin(teta_a),
            cos(gamma_a) * np.exp(1j * phi_a),
            -sin(gamma_a) * cos(teta_a),
        ],
        dtype=np.complex128,
    )
    a_4 = np.array(
        [
            0,
            sin(gamma_a) * cos(teta_a) * np.exp(1j * phi_a) / N_a,
            cos(gamma_a) / N_a,
        ],
        dtype=np.complex128,
    )

    b_0 = np.array(
        [
            0,
            1,
            0,
        ],
        dtype=np.complex128,
    )
    b_1 = np.array(
        [
            sin(gamma_b) * sin(teta_b),
            cos(gamma_b) * np.exp(1j * phi_b),
            -sin(gamma_b) * cos(teta_b),
        ],
        dtype=np.complex128,
    )
    b_2 = np.array(
        [
            1,
            0,
            0,
        ],
        dtype=np.complex128,
    )
    b_3 = np.array(
        [
            cos(teta_b),
            0,
            sin(teta_b),
        ],
        dtype=np.complex128,
    )
    b_4 = np.array(
        [
            0,
            sin(gamma_b) * cos(teta_b) * np.exp(1j * phi_b) / N_b,
            cos(gamma_b) / N_b,
        ],
        dtype=np.complex128,
    )

    alice = [a_0, a_1, a_2, a_3, a_4]
    bob = [b_0, b_1, b_2, b_3, b_4]
    return [np.kron(a, b).reshape(-1, 1) for a, b in zip(alice, bob)]


def parametrized_upb(
    gamma_a: float,
    teta_a: float,
    phi_a: float,
    gamma_b: float,
    teta_b: float,
    phi_b: float,
) -> np.ndarray:
    """The bound entangled state built from the six-parameter family of UPBs.

    Args:
        gamma_a: Alice's gamma angle. Requires cos(gamma_a) != 0 and
            sin(gamma_a) != 0 for the basis to be unextendible.
        teta_a: Alice's theta angle. Requires cos(teta_a) != 0 and
            sin(teta_a) != 0 for the basis to be unextendible.
        phi_a: Alice's phase angle.
        gamma_b: Bob's gamma angle, subject to the same restriction as gamma_a.
        teta_b: Bob's theta angle, subject to the same restriction as teta_a.
        phi_b: Bob's phase angle.

    Returns:
        np.ndarray: the bound entangled state on the orthogonal complement of
        the parametrized unextendible product basis.
    """
    return upb(parametrized_basis(gamma_a, teta_a, phi_a, gamma_b, teta_b, phi_b))

File: js/src/c3OtimesC3/upb/parametrizedUpb.ts
/**
 * https://arxiv.org/abs/quant-ph/9908070
 *
 * D. P. DiVincenzo, T. Mor, P. W. Shor, J. A. Smolin, B. M. Terhal,
 * "Unextendible Product Bases, Uncompletable Product Bases and Bound
 * Entanglement", Commun. Math. Phys. 238, 379 (2003), Section IV A. A
 * six-parameter family of unextendible product bases on C³ ⊗ C³ that contains
 * the Pyramid and Tiles UPBs as special cases.
 */

import { complex, multiply } from 'mathjs';
import type { Matrix } from 'mathjs';
import type { Scalar } from '../../types.js';
import { tensor, upb } from '../../utils/index.js';

export interface ParametrizedUpbOptions {
  /** Alice's gamma angle. Needs `cos(gammaA) != 0` and `sin(gammaA) != 0`. */
  gammaA: number;
  /** Alice's theta angle. Needs `cos(thetaA) != 0` and `sin(thetaA) != 0`. */
  thetaA: number;
  /** Alice's phase angle. */
  phiA: number;
  /** Bob's gamma angle, subject to the same restriction as `gammaA`. */
  gammaB: number;
  /** Bob's theta angle, subject to the same restriction as `thetaA`. */
  thetaB: number;
  /** Bob's phase angle. */
  phiB: number;
}

/** `exp(i·phi)`, the phase the two parties' fourth and fifth vectors carry. */
function phase(phi: number): Scalar {
  return complex(Math.cos(phi), Math.sin(phi));
}

/** One party's five vectors in C³, from that party's three angles. */
function partyVectors(gamma: number, theta: number, phi: number): Scalar[][] {
  const normalization = Math.sqrt(Math.cos(gamma) ** 2 + Math.sin(gamma) ** 2 * Math.cos(theta) ** 2);
  return [
    [1, 0, 0],
    [0, 1, 0],
    [Math.cos(theta), 0, Math.sin(theta)],
    [
      Math.sin(gamma) * Math.sin(theta),
      multiply(Math.cos(gamma), phase(phi)) as Scalar,
      -Math.sin(gamma) * Math.cos(theta),
    ],
    [
      0,
      multiply((Math.sin(gamma) * Math.cos(theta)) / normalization, phase(phi)) as Scalar,
      Math.cos(gamma) / normalization,
    ],
  ];
}

/**
 * The five product vectors of the six-parameter family of UPBs on C³ ⊗ C³.
 *
 * Reduces to the Pyramid UPB for `phiA = phiB = 0` and
 * `thetaA = thetaB = gammaA = gammaB = arccos((√5 - 1)/2)`, and to the Tiles
 * UPB for `phiA = phiB = 0` and `thetaA = thetaB = gammaA = gammaB = 3π/4`.
 *
 * @param options - see {@link ParametrizedUpbOptions}.
 * @returns the five normalized product vectors |a_i> ⊗ |b_i> forming the UPB.
 */
export function parametrizedBasis({
  gammaA,
  thetaA,
  phiA,
  gammaB,
  thetaB,
  phiB,
}: ParametrizedUpbOptions): Matrix[] {
  const alice = partyVectors(gammaA, thetaA, phiA);
  const bob = partyVectors(gammaB, thetaB, phiB);
  // Bob's vectors pair with Alice's in a different order, which is what makes
  // the basis unextendible.
  const bobOrder = [1, 3, 0, 2, 4];
  return alice.map((a, i) => tensor(a, bob[bobOrder[i]]));
}

/**
 * The bound entangled state built from the six-parameter family of UPBs.
 *
 * @param options - see {@link ParametrizedUpbOptions}.
 * @returns the bound entangled state on the orthogonal complement of the UPB.
 */
export function parametrizedUpb(options: ParametrizedUpbOptions): Matrix {
  return upb(parametrizedBasis(options));
}
